---
title: "DE-analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown
### get TCGA COAD miRs
```{r}
####### Network analysis with RTN #########
##to reproduce Fessler data: 
#they combined 30 DE miRs and ~1500 DE mRNAs from CMS4vsCMS1-3 from TCGA-COAD/READ
#BiocManager::install("RTN")
#a) combine DE mRNAs (CMS1vsCMS2-4 + CMS2vs...) from TCGA-COAD with the top RF-miRs and infer network for top RF-miRs?
#b) in CPTAC2 identify CMS4vsCMS1-3 DE mRNA & CMS4vsCMS1-3 DE miR, combine and infer network
#? CMS from miR or CMS from mRNA?

#### get raw mRNA from TCGA ####
## fetch TCGA data COAD:
library(TCGAbiolinks)
clinical.coad <- GDCquery_clinic(project = c("TCGA-COAD"), type = "clinical")
CMS_samples$coad <- clinical.coad$tissue_or_organ_of_origin[match(CMS_samples$SampleId,
                              as.character(clinical.coad$submitter_id))]
oldRowNames <- rownames(read.table("/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/Data/TCGA-COAD-allRNA-01Aonly-RSEM-ComBat.csv",
                                   header = T))
#after packages update, rownames are gone. using the old rownames from a saved table because the number of rows is the same

rownames(clinical.coad) <- clinical.coad$submitter_id
query.exp <- GDCquery(project = c("TCGA-COAD"),
                      legacy = T,
                      data.category = "Gene expression",
                      data.type = "Gene expression quantification",
                      file.type = "results", # or normalized_results
                      experimental.strategy = "RNA-Seq",
                      platform = "Illumina GA"
)
GDCdownload(query.exp, directory = "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/GDCdata")
TCGARnaseqSE <- GDCprepare(query= query.exp, summarizedExperiment = F, save = F,
                           directory = "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/GDCdata")
TCGARnaseqRaw <- round(as.data.frame(TCGARnaseqSE)[,grep("raw_count", colnames(TCGARnaseqSE)) ],0) #or raw_count_
colnames(TCGARnaseqRaw) <- sub("raw_count_","",colnames(TCGARnaseqRaw))
colnames(TCGARnaseqRaw) <- sub("-01A-.*", "",colnames(TCGARnaseqRaw)) 
TCGARnaseqRSEM.GA <- as.data.frame(TCGARnaseqSE)[,grep("scaled_estimate", colnames(TCGARnaseqSE)) ] #or raw_count_
remove(TCGARnaseqSE) #193 for only GA as in Fessler
colnames(TCGARnaseqRSEM.GA) <- sub("-01A-.*", "",sub("scaled_estimate_","",
                                                     colnames(TCGARnaseqRSEM.GA)))
TCGA.COAD.mRna.scale <- scale(TCGARnaseqRSEM.GA, center = F) #scale for RTN
rownames(TCGA.COAD.mRna.scale) <- oldRowNames
rownames(TCGARnaseqRaw) <- oldRowNames

# write.csv(TCGA.COAD.mRna.scale,
#             "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/Data/TCGA-COAD-mRna-scale.csv")
# write.csv(TCGARnaseqRaw,
#             "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/Data/TCGA-COAD-mRna-raw.csv")


## get COAD miRNA from TCGA
## fetch TCGA COAD miRNAs:
library(TCGAbiolinks)
query.exp <- GDCquery(project = c("TCGA-COAD"),
                      legacy = F,
                      data.category = "Transcriptome Profiling",
                      data.type = "miRNA Expression Quantification",
                      experimental.strategy = "miRNA-Seq"
)
GDCdownload(query.exp, directory = "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/GDCdata")
TCGARnaseqSE <- GDCprepare(query= query.exp, summarizedExperiment = F, save = F,
                           directory = "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/GDCdata")
TCGAmiR.COAD <- TCGARnaseqSE[,grep("reads_per_million_", colnames(TCGARnaseqSE)) ] #or read_count_ for raw
colnames(TCGAmiR.COAD) <- sub("reads_per_million_miRNA_mapped_","",colnames(TCGAmiR.COAD))

# ##OR for DE mirs
# TCGAmiR.COAD <- TCGARnaseqSE[,grep("read_count_", colnames(TCGARnaseqSE)) ] #or read_count_ for raw
# colnames(TCGAmiR.COAD) <- sub("read_count_","",colnames(TCGAmiR.COAD))

rownames(TCGAmiR.COAD) <- TCGARnaseqSE$miRNA_ID
remove(TCGARnaseqSE)
##TCGA data has apparently similar counts for isomirs, summarize by mean up
TCGAmiR.COAD$miRcomb <- sub("-[1-9]$","",row.names(TCGAmiR.COAD))
miR_COAD_comb <- as.data.frame(TCGAmiR.COAD %>% dplyr::group_by(miRcomb) %>%
                                 dplyr::summarise_if(.predicate = function(x) is.numeric(x),
                                                     .funs = mean)) #affects around 180miRNA
row.names(miR_COAD_comb) <- miR_COAD_comb$miRcomb #afterwards remove column 1
miR_COAD_comb$miRcomb <- NULL
##keep only one sample per patient
colnames(miR_COAD_comb) <- sub("-01A-.*","",colnames(miR_COAD_comb))
miR_COAD_comb <- miR_COAD_comb[,which(colnames(miR_COAD_comb) %in% 
                                        colnames(TCGA.COAD.mRna.scale))] #no more duplicated patients
##normalize, in right direction for RF
TCGA.COAD.miR.scale <- scale(miR_COAD_comb, center = F)

# write.csv(miR_COAD_comb,
#             "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/Data/TCGA-COAD-GA-miR-raw-counts.csv") #for DEseq
# write.csv(TCGA.COAD.miR.scale,
#             "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/Data/TCGA-COAD-miR-scale.csv")
TCGA.COAD.miR.scale <- read.csv( "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/Data/TCGA-COAD-miR-scale.csv")
miR_COAD_comb <- read.csv( "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/Data/TCGA-COAD-GA-miR-raw-counts.csv", 
                           row.names = 1)

```
### get TCGA COAD GA CMS DE genes
```{r}
###### DE genes for each CMS COAD GA raw mRNAs ######
## DESeq2 getter
getDDS <- function( gene.counts=NULL, phenData=NULL, 
                    mainCond=NULL, listForm = NULL, refList = NULL, rlog = FALSE) {
dds <- DESeqDataSetFromMatrix( countData = gene.counts , 
                                         colData = phenData, 
                                         design = listForm)

  colData( dds )[ , mainCond ] <-  relevel( colData( dds )[ , mainCond ],
                                            ref =  refList )
  colData( dds )[ , mainCond ] <-  droplevels( colData( dds )[ , mainCond ] )

  dds <- dds[ rowSums( counts( dds )) > 1, ] # only keep expressed genes

  if(rlog) {
    dds <- varianceStabilizingTransformation(dds, blind = F ) #blind=TRUE only for unbiased Quality assessment
  }
  return(dds)
}

#library("GSVA")
library("DESeq2")
# browseVignettes("DESeq2")
library("stringi")
library("GenomicFeatures")
library("GenomicAlignments")
library("BiocParallel")
library("Rsamtools")
library("tools")
library("pheatmap")
library("RColorBrewer")
library("ggplot2")
library("gridExtra")
library("genefilter")
library("AnnotationDbi")
library("org.Hs.eg.db")
library(plyr)
library(reshape)
library(ComplexHeatmap)
library(circlize)
library(DT)
library(dplyr)


## directories
geneDB <- org.Hs.eg.db
analysHome <- "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/analyses/DE-results/"
projHome <- "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/analyses/DE-results"
## define Deseq2 analysis param
mainCond <- c( paste0("CMS", 1:4) )
listForm <- list( formula( ~ CMS1 ),
                  formula( ~ CMS2 ),
                  formula( ~ CMS3 ),
                  formula( ~ CMS4 ))
refList <- c( rep("other", 4) )
altHyp <- c(  rep("greaterAbs", 4) )
samples <- c(paste0("CMS", 1:4))
COAD.GA.pat <- colnames(TCGARnaseqRaw)

COAD.GA.pat <- colnames(miR_COAD_comb)
rownames(predictedCMS.COAD) <- gsub("-",".",rownames(predictedCMS.COAD))

grouping <- list( COAD.GA.pat, COAD.GA.pat, COAD.GA.pat, COAD.GA.pat ) 
cmsAnnot <- predictedCMS.COAD[COAD.GA.pat,] #use the LV labels here in order to have labels for all
cmsAnnot$CMS1 <- factor( ifelse(grepl("CMS1", cmsAnnot$LV.table), "CMS1", "other"))
cmsAnnot$CMS2 <- factor( ifelse(grepl("CMS2", cmsAnnot$LV.table), "CMS2", "other"))
cmsAnnot$CMS3 <- factor( ifelse(grepl("CMS3", cmsAnnot$LV.table), "CMS3", "other"))
cmsAnnot$CMS4 <- factor( ifelse(grepl("CMS4", cmsAnnot$LV.table), "CMS4", "other"))


### do DEseq analysis
listDDS <- vector("list", length=length(grouping))
listRLD <- vector("list", length=length(grouping))
listDEDDS <- vector("list", length=length(grouping))
listRes <- vector("list", length=length(grouping))

for( i in 1:length(grouping)) {
  print(paste("now calculating:", samples[i], mainCond[i]), sep=" ")
  listDDS[[i]] <- getDDS( gene.counts = round(miR_COAD_comb[ , grouping[[i]] ],0),#TCGARnaseqRaw[ , grouping[[i]] ], 
                          phenData = cmsAnnot[ grouping[[i]] , ], 
                          mainCond = mainCond[i], 
                          listForm = listForm[[i]],
                          refList = refList[i],
                          rlog = F ) 
  listRLD[[i]] <- getDDS( gene.counts=round(miR_COAD_comb[ , grouping[[i]] ],0),#TCGARnaseqRaw[ , grouping[[i]] ], 
                          phenData = cmsAnnot[ grouping[[i]] ,  ], 
                          mainCond = mainCond[i], 
                          listForm = listForm[[i]],
                          refList = refList[i],
                          rlog = TRUE ) #vst transform with blind=FALSE for mRNA, varianceStabilizingTransformation for miRs
  }
listDEDDS <- lapply( listDDS, DESeq )
#listDEDDS[[i]] <- DESeq(listDDS[[i]])
for( i in 1:length(grouping) ) {
  listRes[[i]] <- results(listDEDDS[[ i ]], 
           altHypothesis = altHyp[ i ],
           alpha=0.05)
}
lapply( listRes, summary ) 
summary(listRes[[1]])
minAdjP <- 0.001
minAdjP <- 0.05 #for miRs??

##### INSPECT RESULTS #####
i=1
DEgenes <- data.frame("CMS"=NA,
                      "GeneID"=NA,
                      "padj"=NA,
                      "FC"=NA,
                      "meanB"=NA)
DEgenes200 <- data.frame("CMS"=NA,
                      "GeneID"=NA,
                      "padj"=NA,
                      "FC"=NA,
                      "meanB"=NA)
for( i in 1:length(grouping)) {
  ## which genes are DE in one or another condition ##
  res <- listRes[[i]]
  rld <- listRLD[[i]]
  normcounts <- data.frame( round( counts(listDEDDS[[i]], normalized=TRUE), 
                                   digits = 1 ) )
  
  ## add more info to results table ##
  res$symbol <- sub("\\|.*","",row.names(res))
  res$entrez <- sub(".*\\|","",row.names(res))
  columnNames <- colnames(res)
  res <- cbind( res, normcounts )
  colnames( res ) <- c(columnNames, colnames(assay(rld)))

  ## write results table ##
  resOrdered <- as(res[ order(res$padj, -abs(res$stat), na.last = T ), ],
                   "data.frame")
  resOrderedSig <- resOrdered[ which( resOrdered$padj < minAdjP & # for mRNA adjust to Fessler paper: 0.001 and log2FC 0.85
                                        abs(resOrdered$log2FoldChange) > abs(log2(0.71))), ] # for miR according to Fessler paper: FDR 0.001 and FC < 0.71

  write.csv( data.frame( "GeneID"=row.names(resOrderedSig), resOrderedSig ),
             file = paste0( analysHome, "/resSigTable-p0.05fc0.71-COAD-GA-miRs-LV-CMS-",
                            samples[i], 
                            ".csv"),
              row.names = F )

  res4GSEA <- resOrdered[ order( resOrdered$stat, resOrdered$log2FoldChange), ] # order increasing -> rank will be used in a decreasing manner in GSEA-Preranked
  write.table( cbind( res4GSEA$symbol, res4GSEA$stat ),
               file = paste0( analysHome, "/rankGSEA-miRs-",
                               mainCond[i], #find useful file name
                              "-stat.rnk"),
               row.names = F, sep = "\t", col.names = F, quote = F)
  
  DEgenes<- rbind(na.exclude(DEgenes), data.frame("CMS"=paste0("CMS", i),
                            "GeneID"=rownames(resOrderedSig),
                            "padj"=resOrderedSig$padj,
                            "stat"=resOrderedSig$stat, 
                            "FC"=resOrderedSig$log2FoldChange, 
                            "meanB"=resOrderedSig$baseMean))
  DEgenes200 <- rbind(na.exclude(DEgenes200), data.frame("CMS"=paste0("CMS", i),
                            "GeneID"=rownames(resOrderedSig)[1:200],
                            "padj"=resOrderedSig$padj[1:200],
                            "stat"=resOrderedSig$stat[1:200], 
                            "FC"=resOrderedSig$log2FoldChange[1:200], 
                            "meanB"=resOrderedSig$baseMean[1:200]))
}
DEgenes <- DEgenes[order(DEgenes$padj, decreasing = F),]
summary(DEgenes[which(duplicated(DEgenes$GeneID)),"CMS"]) # any genes DE in more than one CMS?
#less overlap with CMS4
mRNAs.CMS <-  unique(c(as.character(DEgenes[which(duplicated(DEgenes$GeneID)),"GeneID"]), 
                       as.character(DEgenes200$GeneID ) ) )#1800
#selection of interesting genes

write.csv(DEgenes200,
            "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/analyses/TCGA-COAD-GA-mRna-CMS-DEgenes200.csv")
write.csv(DEgenes200,
            "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/analyses/TCGA-COAD-GA-miRs-CMS-DEgenes200.csv")

```
### OR get CPTAC2 CRC miRs 
```{R}
###### CPTAC-2: has mir-Seq and CMS from mRNA-Seq, plus clinical data, mutations etc. 
###### but  has no survival data

query.exp <- GDCquery(project = c("CPTAC-2"), 
                      data.category = "Transcriptome Profiling",
                      data.type = "miRNA Expression Quantification", 
                      experimental.strategy = "miRNA-Seq"
)
GDCdownload(query.exp, directory = "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/GDCdata")
getResults(query.exp )
TCGARnaseqSE <- GDCprepare(query= query.exp, summarizedExperiment = F, save = F,
                           directory = "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/GDCdata")
TCGAmiR.Cptac2 <- TCGARnaseqSE[,grep("reads_per_million", colnames(TCGARnaseqSE)) ] #or reads_per_million for normalized_results
colnames(TCGAmiR.Cptac2) <- sub("reads_per_million_miRNA_mapped","",colnames(TCGAmiR.Cptac2))
rownames(TCGAmiR.Cptac2) <- TCGARnaseqSE$miRNA_ID
remove(TCGARnaseqSE)
##TCGA data has apparently similar counts for isomirs, summarize by mean up
TCGAmiR.Cptac2$miRcomb <- sub("-[1-9]$","",row.names(TCGAmiR.Cptac2))
miR_Cptac_comb <- as.data.frame(TCGAmiR.Cptac2 %>% dplyr::group_by(miRcomb) %>% 
                                 dplyr::summarise_if(.predicate = function(x) is.numeric(x),
                                                     .funs = mean)) #affects around 180miRNA
row.names(miR_Cptac_comb) <- miR_Cptac_comb$miRcomb #afterwards remove column 1
miR_Cptac_comb$miRcomb <- NULL
##normalize, in right direction for RF
CPTAC2.miR.scale <- scale( miR_Cptac_comb[, 
                                   which( cptac.ids$subm %in% clinSupp.cptac$SampleID ) ],
                           center=F) ##only the subset from the publication suppl clin data
rownames(CPTAC2.miR.scale) <- gsub("-", ".", rownames(CPTAC2.miR.scale)) # format mir names for RF

# write.csv(CPTAC2.miR.scale,
#           "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/Data/TCGA-CPTAC2.miR.RPM.scale.csv")

## clinical data CPTAC1 from GDC or from publication supplement
clinical.cptac <- GDCquery_clinic(project = c("CPTAC-2"), type = "clinical")

clinSupp.cptac <- read.table("/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/Data/Vasaikar_CPTAC2_STable1_mmc1.txt", 
                             sep="\t", header = T)
clinSupp.cptac$sample.label <- cptac.ids$coln[match(clinSupp.cptac$SampleID,
                                                    cptac.ids$subm)]
rownames(clinSupp.cptac) <- clinSupp.cptac$SampleID

##identify how submitter id and sample id map
cptac.ids <- data.frame("coln"=colnames(miR_Cptac_comb),
                        "subm"=NA)
for(i in 1:length(colnames(miR_Cptac_comb))){
cptac.ids$subm[i] <- as.character(clinical.cptac$submitter_id[ grep(colnames(miR_Cptac_comb)[i], 
     as.character(clinical.cptac$submitter_sample_ids) ) ])
}
colnames(miR_Cptac_comb)==cptac.ids$coln



##get also the mRNA data from CPTAC
query.exp <- GDCquery(project = c("CPTAC-2"), 
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification", 
                      experimental.strategy = "RNA-Seq",
                      workflow.type = "HTSeq - FPKM"#??
)
GDCdownload(query.exp)
getResults(query.exp )
CPTAC2.mRna.FPKM <- GDCprepare(query= query.exp, summarizedExperiment = F, save = F)
genesCPTAC2 <- CPTAC2.mRna.FPKM$X1
symbolsCPTAC2 <- mapIds(org.Hs.eg.db, keys=sub("\\..*","",genesCPTAC2), 
                     'SYMBOL', 'ENSEMBL')
##normalized FPKM, just scale for RTN
CPTAC2.mRna.FPKM <- scale( CPTAC2.mRna.FPKM[, which( colnames(CPTAC2.mRna.FPKM)
                                                     %in% 
                                                       gsub("\\.", "-", 
                                                            colnames( miR_Cptac_comb) ) )],
                           center=F) ##only the subset from the publication suppl clin data
rownames(CPTAC2.mRna.FPKM) <- unname(symbolsCPTAC2)



```

## Network
```{r}
######## RTN network ########
library(RTN)
library(DESeq2)
library(RedeR)

#### top 20 miR from RF (50x, same for 80x) ####
tfs <- c("hsa-mir-625", "hsa-mir-592", "hsa-mir-552", "hsa-mir-218" ,
         "hsa-mir-31", "hsa-mir-375", "hsa-mir-143",  "hsa-mir-615",  
         "hsa-mir-335",  "hsa-mir-146b", "hsa-mir-99a", "hsa-mir-141", 
         "hsa-mir-92b" ,  "hsa-mir-942", "hsa-mir-3170",  "hsa-mir-362",  
         "hsa-mir-30a",   "hsa-mir-155", "hsa-mir-582", "hsa-mir-92a" )
#### just in case WNT genes need anotation ####
markerGenesFlierWNTup_AdCa <- c("Cemip","(Kiaa1199)", "Foxq1", "Cldn2", "Etv4", "Phlda1", "Hes6", "Cd44", "Slc7a5", "Sord", "Rnf43", "(Flj20315)", "Ctps", "Enc1", "Tead4", "Met", "Myc", "Rrp12", "(Kiaa0690)", "C19orf48","(Mgc13170)", "Tspan5","(Tm4sf9)", "Cdc25a", "Cdkn3", "Bysl", "Scd", "Sox9","Pdcd2l", "(Mgc13096)", "Sox4", "Wdr77", "Trmt1", "Nob1p", "Psf1", "Nle1", "Pmscl1", "Eif5a2", "Tgif", "Ppil1", "Ets2", "Slc39a10", "Rrp9", "(Rnu3ip2)", "Utp4","(Cirh1a)", "Znf511", "Ddx21", "Dcun1d5", "Hilpda", "Znf593", "Mrto4", "Dkc1", "Odc1", "Cdk4","Psmg4", "(Loc389362)", "Ppan", "Mac30","Cytor", "(Mgc4677)", "Nop16", "Aurkb", "C12orf45","(Mgc40397)", "Cd320", "Znf703", "Wdr12", "Ns", "Cdt1", "Pno1","(Loc56902)", "Sdccag16", "Noc3l", "Peo1", "Cad", "Xpot", "Lyar","(Flj20425)", "Polr1c", "Nat10","(Flj10774)", "Cdca7", "Wdr74", "Yap1", "Mtvr1", "Srm", "Nxt1", "Snhg15","(Loc285958)", "Ece2","(Mgc2408)", "Mrpl36", "Ddx10", "Mrps12", "Rabepk", "(Rab9p40)", "Rheb", "Wdr3", "Rfp", "Polr1b", "Gpatch4", "(Flj20249)", "Ddx20", "Nifk","(Mki67ip)", "Farsa", "Crtap", "Sf3a2", "Gemin5", "Polr1d", "Mettl1", "Rrp1b","(Kiaa0179)", "Znf259", "Rrp41", "Hnrpdl", "Nprl5","(C16orf35)", "Ccdc86", "(Mgc2574)", "Ddx56", "Trex2", "Nola1","Rrp7a", "(Cgi-96)", "Imp4", "Mrps30", "Yrdc", "(Flj23476)", "Heatr1", "Slc29a2", "Tere1", "Mrps26", "Slc19a1", "Ydjc","(Loc150223)", "Tp53rk", "(C20orf64)", "Cog8","Fam60a", "(C12orf14)", "Mybbp1a", "Pdcd11", "Nop2", "Nol5a", "Slc3a2", "Nudcd3", "(Kiaa1068)")
markerGenesFlierWNTup_onlyCa <- c("CXCL5", "ZIC2", "CBX2",  "C20orf82", "GLS2", "BAG2", "ZNF503", "SNHG17","(LOC388796)", "LEF1", "FLJ12683", "", "NELF", "HOXA11", "CKLFSF7", "LIMS1", "SLC19A2", "SCLY", "HIRA", "SLC52A2","(GPR172A)", "DOCK4", "AMPD2", "MAT2A", "C1orf109","(FLJ20508)", "TPST2", "GASK1B", "(DKFZp434L142)", "PPIF", "DUS3L","(LOC56931)", "ISG20L2","(FLJ12671)", "SLC25A19", "SLC1A5", "TIMM10", "POLRMT", "ITGB4BP", "ZNF275", "SFRS7", "RBM12", "UMPK")
markerGenesFlierWNTup_onlyAd <- c("Ascl2", "Axin2", "Gpr49", "Tdgf1", "Tbx3", "Ephb3", "Ephb2", "Znrf3", "Ucc1", "Il20ra", "Rhobtb3", "Slc5a1", "STAMBPL1","(Amsh-Lp)", "Gemin4", "Gpx2", "Id3", "Skb1", "Kcne3", "CEP76","(C18orf9)","RRP15", "(Cgi-115)", "ZNHIT2", "(C11orf5)", "Nse1", "Fut4", "Dhx33", "KIZ", "(C20orf19)", "Gemin6", "Jag1", "Pigw", "Myb", "Apex1", "SPIN4", "(Loc139886)", "NIP7", "(Cgi-37)", "Hsa9761", "SYBU","(Flj20366)", "Ptma", "Mgc16824", "Hspc242", "Znf278",  "PAAF1", "(Flj11848)", "Srprb", "Znrd1", "Tfb2m", "Rpc62", "Arid5b", "Hnrpa0", "Mettl2", "Jtv1", "Loc20525", "Mrpl15", "Znf339", "Mrpl12")
wntGenes <- toupper( c(markerGenesFlierWNTup_AdCa, markerGenesFlierWNTup_onlyCa, markerGenesFlierWNTup_onlyAd))

### get the data ####
TCGA.COAD.mRna.scale <- read.csv("/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/Data/TCGA-COAD-mRna-scale.csv", 
                                 row.names=1)

TCGA.COAD.miR.scale <- read.csv("/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/Data/TCGA-COAD-miR-scale.csv", 
                                 row.names=1)
rownames(TCGA.COAD.miR.scale) <- gsub("-",".", rownames(TCGA.COAD.miR.scale))


DEgenes <- read.csv("/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/analyses/DE-results/resSigTable-COAD-GA-mRNA-LV-CMS-CMS1.csv",
                      as.is = T)

DEgenes <- DEgenes[which( DEgenes$GeneID %in% rownames(TCGA.COAD.mRna.scale)), ]


DEmiRS <- read.csv("/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/analyses/DE-results/resSigTable-p0.05fc0.71-COAD-GA-miRs-LV-CMS-CMS3.csv",
                      as.is = T)

DEmiRS$GeneID <- gsub("-",".", DEmiRS$GeneID)
DEmiRS <- DEmiRS[which( DEmiRS$GeneID %in% rownames(TCGA.COAD.miR.scale)), ]

##identified 200 top DE genes for each CMS and |log2FC|>0.85, FDR<0.001
##from TCGA-COAD-GA
#### merge data according to RTN vignette ####
expData <- rbind( TCGA.COAD.miR.scale[ as.character( DEmiRS$GeneID ) ,
                                       colnames(TCGA.COAD.miR.scale)],
                  TCGA.COAD.mRna.scale[ as.character( DEgenes$GeneID ) ,
                                       colnames(TCGA.COAD.miR.scale)])
rowAnnotation <- data.frame("PROBEID"=rownames(expData), 
                           "SYMBOL"=sub("\\|.*","", rownames(expData)),
                            "WNT"=as.numeric( sub("\\|.*","", rownames(expData)) %in% wntGenes ),
                           "L2FC"=round( c( DEmiRS$log2FoldChange, DEgenes$log2FoldChange), 1),
                            "NRF"=round( importances_df[match(rownames(expData), 
                                                             rownames( importances_df)),"Overall"]) )
rowAnnotation$NRF[is.na(rowAnnotation$NRF)] <- 0
colAnnot <- cbind( data.frame("ID"=rownames(cmsAnnot[colnames(TCGA.COAD.miR.scale),]) ),
                   cmsAnnot[colnames(TCGA.COAD.miR.scale),])
colAnnot$LV.table[is.na(colAnnot$LV.table)] <- "NOLBL"


#### RTN analysis ####
rtni <- tni.constructor(expData = as.matrix(expData), 
                        regulatoryElements = DEmiRS$GeneID[which(DEmiRS$pval<0.001 & #this is according to Fessler paper
                                                                DEmiRS$log2FoldChange > 0)], #check the most DE miRs
                        ## also plot upregulated miRs but separately from downregualted miRs  
                        rowAnnotation = rowAnnotation, 
                        colAnnotation = colAnnot)
rtni <- tni.permutation(rtni)
rtni <- tni.bootstrap(rtni)
rtni <- tni.dpi.filter(rtni, eps = NA) 

#regulons <- tni.get(rtni, what = "regulons.and.mode", idkey = "SYMBOL")
# write.csv(data.frame(names(regulons$hsa.mir.141), regulons$hsa.mir.141 ), "/Users/ronjaadam/projects/miRNA_mCRC/miRNA classifier/analyses/Network-RTN/TCGA-COAD-GA-CMS-DEgenes-all-CMS4-miR141-regulon.txt",
#           row.names = F, quote = F)

#### RTN graph via RedeR ####
g <- tni.graph(rtni) #default parameters got CMS4 down close to Fessler Figure 
g <- att.setv(g, from = "L2FC", to = "nodeColor", breaks=seq(-5,5,0.8), pal=2)
g <- att.setv(g, from = "NRF", to = "nodeSize")
rdp <- RedPort()
calld(rdp, checkcalls=TRUE)
resetd(rdp)
addGraph(rdp, g, layout=NULL)
addLegend.color(rdp, g, type="edge")
addLegend.color(rdp, g, type="node")
addLegend.size(rdp, g, type="node")
addLegend.shape(rdp, g)
relax(rdp, ps = TRUE)



# ##CPTAC2 #tried but network was quite empty due to the low miR expression
# for (i in c(1, 201, 401, 601) ){ 
# i=401
# cmsGenesValid <- which(sub("\\|.*","",DEgenes200$GeneID[i:(i+199)]) %in% rownames(CPTAC2.mRna.FPKM))
# cmsSymbols <- sub("\\|.*","", DEgenes200$GeneID[cmsGenesValid+i-1])
# expData <- rbind( CPTAC2.miR.scale[ rownames( importances_df[1:30,]),],
#                   CPTAC2.mRna.FPKM[cmsSymbols, match(colnames(CPTAC2.miR.scale), 
#                                                      colnames(CPTAC2.mRna.FPKM))])
# rowAnnotation <- data.frame("PROBEID"= rownames(expData), 
#                             "SYMBOL"= rownames(expData) )
# rownames(rowAnnotation) <- rownames(expData)
# 
# colAnnot <- data.frame("ID"=clinSupp.cptac[match(colnames(CPTAC2.miR.scale), 
#                                                         clinSupp.cptac$sample.label),"sample.label"] ,
#                    "CMS"=factor(clinSupp.cptac[match(colnames(CPTAC2.miR.scale), 
#                                                         clinSupp.cptac$sample.label),"CMS"],
#                                 levels=c("CMS1", "CMS2", "CMS3", "CMS4", "NOLBL")))
# colAnnot$CMS[is.na(colAnnot$CMS)] <- "NOLBL"
# View(miR_Cptac_comb[ gsub("\\.","-", rownames( importances_df[1:20,])), 
#                      order( clinSupp.cptac[match(colnames(miR_Cptac_comb), 
#                                                         clinSupp.cptac$sample.label),"CMS"])  ])
# ##plot density distribution
# miR_Cptac_comb_t <- data.frame( log2( t(miR_Cptac_comb) +1))
# miR_Cptac_comb_t$CMS <- clinSupp.cptac[match(rownames(miR_Cptac_comb_t), 
#                                              clinSupp.cptac$sample.label),"CMS"]
# 
# ggplot( miR_Cptac_comb_t, aes( x= hsa.mir.92a) ) + 
#   geom_density( aes( group=CMS, colour=CMS, fill=CMS ), alpha=0.3 ) +
#   scale_fill_manual( values = paletteCMS ) +
#   scale_color_manual( values = paletteCMS ) +
#   theme_minimal( )
#}

```


Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
